src/lib/raft/raft.c

```diff
@@ -1211,6 +1211,7 @@ raft_promote(struct raft *raft)
 		return;
 	raft_sm_schedule_new_term(raft, raft->volatile_term + 1);
 	raft_start_candidate(raft);
+	raft_sm_schedule_new_vote(raft, raft->self, raft->vclock);
```

Comment:
Not really. We have to do `raft_sm_schedule_new_term` before `raft_start_candidate` for the sake of 'manual' nodes. Otherwise `raft_sm_schedule_new_term` would restore a manual node to non-candidate again.
At the same time we have to do `raft_sm_schedule_new_vote` after `raft_start_candidate` (again, for the manual nodes).

On the other hand I can do this, because `raft_start_candidate` is a no-op for a real candidate. I'm not sure it's worth it though. What do you think? cc: @Gerold103 
```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index f31cee5b1..e0b073b5c 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -959,8 +959,7 @@ raft_sm_schedule_new_election(struct raft *raft)
        say_info("RAFT: begin new election round");
        assert(raft->is_cfg_candidate);
        /* Everyone is a follower until its vote for self is persisted. */
-       raft_sm_schedule_new_term(raft, raft->volatile_term + 1);
-       raft_sm_schedule_new_vote(raft, raft->self, raft->vclock);
+       raft_promote(raft);
 }
 
 static void
```

Reply:
Thank you for explanation, I see now, that `raft_start_candidate` has to be exactly between term and vote when manual promote is called. And I would rather not apply change you proposed, as indeed `raft_start_candidate` won't have any effect in automatic elections, and it would make the code even harder to understand, I'm on the side of explicit call of `new_term` and `new_vote` here.

Reply:
Ok, I don't like this hunk either. Let's revert it then:
```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index e0b073b5c..f31cee5b1 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -959,7 +959,8 @@ raft_sm_schedule_new_election(struct raft *raft)
        say_info("RAFT: begin new election round");
        assert(raft->is_cfg_candidate);
        /* Everyone is a follower until its vote for self is persisted. */
-       raft_promote(raft);
+       raft_sm_schedule_new_term(raft, raft->volatile_term + 1);
+       raft_sm_schedule_new_vote(raft, raft->self, raft->vclock);
 }
 
 static void
```